scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua

```diff
@@ -3,117 +3,534 @@
 --  NPC: Squintrox Dryeyes
 -- Type: Addon Mission Merchant
 -- !pos -100.071 -1 11.869 246
+-- wiki has good info on his behavior: https://ffxiclopedia.fandom.com/wiki/Squintrox_Dryeyes
+-- Essentially, he accomodates re-obtaining key items from completed mini expansion missions
 -----------------------------------
 local ID = zones[xi.zone.PORT_JEUNO]
+require('scripts/missions/amk/helpers')
 -----------------------------------
 local entity = {}
 
 entity.onTrade = function(player, npc, trade)
-    local count         = trade:getItemCount()
-    local sLux          = trade:hasItemQty(xi.item.SEEDSPALL_LUX, 1)
-    local sLuna         = trade:hasItemQty(xi.item.SEEDSPALL_LUNA, 1)
-    local sAstrum       = trade:hasItemQty(xi.item.SEEDSPALL_ASTRUM, 1)
-    local acpMission    = player:getCurrentMission(xi.mission.log_id.ACP)
-    local hasCrimsonKey = player:hasKeyItem(xi.ki.CRIMSON_KEY)
-    local lastCrimson   = player:getCharVar('LastCrimsonKey') -- When last Crimson key was obtained
+    local now                   = os.time()
+    local nextMidnight          = getMidnight()
+    local count                 = trade:getItemCount()
+    local finishedACP           = player:getCurrentMission(xi.mission.log_id.ACP) == xi.mission.id.acp.A_CRYSTALLINE_PROPHECY_FIN
+    local finishedAMK           = player:getCurrentMission(xi.mission.log_id.AMK) == xi.mission.id.amk.A_MOOGLE_KUPO_DETAT_FIN
+    local finishedASA           = player:getCurrentMission(xi.mission.log_id.ASA) == xi.mission.id.asa.A_SHANTOTTO_ASCENSION_FIN
+    local hasMoogleKey          = player:hasKeyItem(xi.ki.MOOGLE_KEY)
+    local kit                   = player:getCharVar('ASA_kit')
+    local hasMoogleItems        = kit ~= 0 and
+                                    trade:hasItemQty(kit, 1) and
+                                    count == 1
+    local hasCrimsonItems       = trade:hasItemQty(xi.item.SEEDSPALL_LUX, 1)        and
+                                    trade:hasItemQty(xi.item.SEEDSPALL_LUNA, 1)     and
+                                    trade:hasItemQty(xi.item.SEEDSPALL_ASTRUM, 1)   and
+                                    count == 3
+    local hasWhiteCoralItems    = trade:hasItemQty(xi.item.ORCISH_PLATE_ARMOR, 1)   and
+                                    trade:hasItemQty(xi.item.QUADAV_BACKSCALE, 1)   and
+                                    trade:hasItemQty(xi.item.YAGUDO_CAULK, 1)       and
```

Comment:
This giant block of things defined at the top is quite unsightly, but this is quite an overloaded NPC - so the option are to use these blocks inline if they're only used one (like here), or to break them out into local functions that takes the trade and/or the player as args and returns a bool

Reply:
Figured it was better to define all the complex requirements in one place at the top, but I could put them directly in the if blocks if that's preferred